esque/cli/commands/produce.py

```diff
@@ -96,55 +108,60 @@ def produce(
-    if topic not in map(attrgetter("name"), topic_controller.list_topics(get_topic_objects=False)):
-        click.echo(f"Topic {blue_bold(topic)} does not exist in context {blue_bold(to_context)}.")
-        if ensure_approval("Would you like to create it now?"):
+    if not read_from_stdin:
+        if not directory:
+            raise ValueError("Need to provide directory if not reading from stdin.")
+        else:
+            directory = pathlib.Path(directory)
+    elif avro:
+        raise ValueError("Cannot read avro data from stdin. Use a directory instead.")
+
+    if binary and avro:
+        raise ValueError("Cannot set data to be interpreted as binary AND avro.")
+
+    topic_controller = Cluster().topic_controller
+    if not topic_controller.topic_exists(topic):
+        if ensure_approval(f"Topic {topic!r} does not exist, do you want to create it?", no_verify=state.no_verify):
             topic_controller.create_topics([Topic(topic)])
         else:
-            raise TopicDoesNotExistException(f"Topic {topic} does not exist!", -1)
-
-    stdin = click.get_text_stream("stdin")
-    if read_from_stdin and isatty(stdin):
-        click.echo(
-            "Type the messages to produce, "
-            + ("in JSON format, " if not ignore_stdin_errors else "")
-            + blue_bold("one per line")
-            + ". End with "
-            + blue_bold("CTRL+D")
-            + "."
-        )
-    elif read_from_stdin and not isatty(stdin):
-        click.echo(f"Reading messages from an external source, {blue_bold('one per line')}).")
-    else:
-        click.echo(
-            f"Producing from directory {blue_bold(str(directory))} to topic {blue_bold(topic)}"
-            f" in target context {blue_bold(to_context)}"
-        )
-    producer = ProducerFactory().create_producer(
-        topic_name=topic,
-        input_directory=input_directory if not read_from_stdin else None,
-        avro=avro,
-        match=match,
-        ignore_stdin_errors=ignore_stdin_errors,
-        binary=binary,
-    )
-    total_number_of_messages_produced = producer.produce()
+            click.echo(click.style("Aborted!", bg="red"))
+            return
+
+    builder = PipelineBuilder()
+
+    input_handler = create_input_handler(directory, read_from_stdin)
+    builder.with_input_handler(input_handler)
+
+    input_message_serializer = create_input_message_serializer(directory, avro, binary)
+    builder.with_input_message_serializer(input_message_serializer)
+
+    output_message_serializer = create_output_serializer(avro, binary, topic, state)
+    builder.with_output_message_serializer(output_message_serializer)
+
+    output_handler = create_output_handler(to_context, topic)
+    builder.with_output_handler(output_handler)
+
+    total_number_of_messages_produced = 0
+
+    if match:
+        builder.with_stream_decorator(yield_only_matching_messages(match))
+
+    def counter_decorator(message_stream: MessageStream) -> MessageStream:
```

Comment:
This looks like the same decorator from the consume command? Since we'll probably need it for transfer as well, how about extracting it?

Reply:
Yeah, good point :+1: